// openapicmd/lib/__tests__/test-fixtures.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRequestBody = exports.createParameter = exports.createOperation = exports.createDefinition = void 0;
const createDefinition = (overrides) => {
    return {
        openapi: '3.0.0',
        servers: [],
        paths: {},
        ...overrides,
        info: {
            title: 'My API',
            version: '0.0.1',
            ...overrides === null || overrides === void 0 ? void 0 : overrides.info
        },
    };
};
exports.createDefinition = createDefinition;
const createOperation = (overrides) => {
    return {
        operationId: 'operationId',
        responses: {},
        ...overrides,
    };
};
exports.createOperation = createOperation;
const createParameter = (overrides) => {
    return {
        name: 'name',
        in: 'query',
        ...overrides,
    };
};
exports.createParameter = createParameter;
const createRequestBody = (overrides) => {
    return {
        content: {},
        ...overrides,
    };
};
exports.createRequestBody = createRequestBody;
